theano/gpuarray/opt.py

```diff
@@ -319,7 +319,8 @@ def apply(self, fgraph):
 
             context_name = None
             for i in [mapping[i] for i in node.inputs]:
-                if isinstance(i.type, GpuArrayType):
+                target = getattr(i.tag, 'target', None)
+                if (isinstance(i.type, GpuArrayType) and target != 'cpu'):
```

Comment:
@abergeron is this the right way to handle in the middle of the graph the target? I don't find the code that support that in the current code in gpuarray/opt.py. Do we support it now? @lamblin told me so. I want this in this PR only if we already supported that. If we didn't supported that, then we should do that in another PR.